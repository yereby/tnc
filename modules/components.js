export function weaponsComp(list, playAction) {
  const ul = document.createElement('ul')
  ul.classList.add('weaponsList')

  list.map((weapon) => {
    const newLi = document.createElement('li')

    const newWeapon = document.createElement('img')
    newWeapon.src = `/pics/${weapon}.png`
    newWeapon.classList.add('weapon')
    newWeapon.alt = weapon
    newWeapon.width = '100'
    newWeapon.onclick = () => playAction(weapon)

    newLi.appendChild(newWeapon)
    ul.appendChild(newLi)
  })

  return ul
}

export function imageComp(image) {
  const pic = document.createElement('img')
  pic.src = `./pics/${image}.png`
  pic.alt = image
  pic.width = '100'

  return pic
}

/*
 * Create the winners component
 * We display a message depending of who is the winner
 * and then we create the weapons list
 *
 * @param {object } res - Result of the game
 * @param {boolean} isPlayer - Human or not
 * @param {function} spock - The method for being spocked
 */
export function winnersComp({ firstWeapon, secondWeapon, winner }, isPlayer, spock) {
  const title = document.createElement('p')
  if (winner === 'firstWeapon') {
    title.innerHTML = isPlayer ? 'You <b>win</b>.' : `<b>${firstWeapon}</b> wins.`
  }
  if (winner === 'secondWeapon') {
    title.innerHTML = isPlayer ? 'You <b>lose</b>.' : `<b>${secondWeapon}</b> wins.`
  }
  if (winner === null) {
    title.innerHTML = 'Draw.'
    if (firstWeapon === 'spock') { spock(true) }
  }

  const ul = document.createElement('ul')
  ul.classList.add('weaponsList')

  const firstPic = imageComp(firstWeapon)
  const secondPic = imageComp(secondWeapon)

  for (let i = 0; i < 3; i++) {
    const newLi = document.createElement('li')
    if (i === 0 || i === 2) { newLi.appendChild(i === 0 ? firstPic : secondPic) }
    if (i === 1) { newLi.innerHTML = 'VS.' }

    ul.appendChild(newLi)
  }

  const list = document.createElement('div')
  list.classList.add('resultList')
  list.appendChild(title)
  list.appendChild(ul)

  return list
}

export function spockedComp() {
  const spocked = document.createElement('img')
  spocked.src = './pics/spocked.jpg'
  spocked.classList.add('spocked')
  spocked.alt = 'You have been Spocked'
  spocked.title = 'You have been Spocked'

  return spocked
}
