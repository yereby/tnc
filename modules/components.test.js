import { weaponsComp, imageComp, spockedComp, winnersComp } from './components.js'

test('Weapons creation', () => {
  const list = weaponsComp(['firstWeap', 'secondWeap'])

  const actual = list.children
  const expectedLength = 2
  const actualChild = list.firstElementChild
  const expectedChildTag = 'LI'
  const actualWeapon = actualChild.firstElementChild
  const expectWeaponClass = 'weapon'

  expect(actual).toHaveLength(expectedLength)
  expect(actualChild.tagName).toBe(expectedChildTag)
  expect(actualWeapon.classList).toContain(expectWeaponClass)
})

test('Weapons creation snapshot', () => {
  const list = weaponsComp(['firstWeap', 'secondWeap'])
  expect(list).toMatchSnapshot()
})

test('Weapons action', () => {
  const list = weaponsComp(['firstWeap'], () => 'Hello')

  const actual = list.firstElementChild.firstElementChild.onclick()
  const expected = 'Hello'

  expect(actual).toBe(expected)
})

test('First winner creation snapshot for a player', () => {
  const res = {
    firstWeapon: 'scissors',
    secondWeapon: 'paper',
    winner: 'firstWeapon',
  }

  const list = winnersComp(res, true, () => 'Hello')
  expect(list).toMatchSnapshot()
})

test('First winner creation snapshot', () => {
  const res = {
    firstWeapon: 'scissors',
    secondWeapon: 'paper',
    winner: 'firstWeapon',
  }

  const list = winnersComp(res, false, () => 'Hello')
  expect(list).toMatchSnapshot()
})

test('Second winner creation snapshot for a player', () => {
  const res = {
    firstWeapon: 'paper',
    secondWeapon: 'scissors',
    winner: 'secondWeapon',
  }

  const list = winnersComp(res, true, () => 'Hello')
  expect(list).toMatchSnapshot()
})

test('Second winner creation snapshot', () => {
  const res = {
    firstWeapon: 'paper',
    secondWeapon: 'scissors',
    winner: 'secondWeapon',
  }

  const list = winnersComp(res, false, () => 'Hello')
  expect(list).toMatchSnapshot()
})

test('Draw game creation snapshot', () => {
  const res = {
    firstWeapon: 'paper',
    secondWeapon: 'scissors',
    winner: null,
  }

  const list = winnersComp(res, false, () => 'Hello')
  expect(list).toMatchSnapshot()
})

test('Draw game creation snapshot with spock', () => {
  const res = {
    firstWeapon: 'spock',
    secondWeapon: 'spock',
    winner: null,
  }

  const list = winnersComp(res, false, () => 'Hello')
  expect(list).toMatchSnapshot()
})

test('Image creation snapshot', () => {
  const img = imageComp()
  expect(img).toMatchSnapshot()
})

test('Spock snapshot', () => {
  const spocked = spockedComp()
  expect(spocked).toMatchSnapshot()
})
