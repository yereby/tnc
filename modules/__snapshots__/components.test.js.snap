// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draw game creation snapshot 1`] = `
<div
  class="resultList"
>
  <p>
    Draw.
  </p>
  <ul
    class="weaponsList"
  >
    <li>
      <img
        alt="paper"
        src="./pics/paper.png"
        width="100"
      />
    </li>
    <li>
      VS.
    </li>
    <li>
      <img
        alt="scissors"
        src="./pics/scissors.png"
        width="100"
      />
    </li>
  </ul>
</div>
`;

exports[`Draw game creation snapshot with spock 1`] = `
<div
  class="resultList"
>
  <p>
    Draw.
  </p>
  <ul
    class="weaponsList"
  >
    <li>
      <img
        alt="spock"
        src="./pics/spock.png"
        width="100"
      />
    </li>
    <li>
      VS.
    </li>
    <li>
      <img
        alt="spock"
        src="./pics/spock.png"
        width="100"
      />
    </li>
  </ul>
</div>
`;

exports[`First winner creation snapshot 1`] = `
<div
  class="resultList"
>
  <p>
    <b>
      scissors
    </b>
     wins.
  </p>
  <ul
    class="weaponsList"
  >
    <li>
      <img
        alt="scissors"
        src="./pics/scissors.png"
        width="100"
      />
    </li>
    <li>
      VS.
    </li>
    <li>
      <img
        alt="paper"
        src="./pics/paper.png"
        width="100"
      />
    </li>
  </ul>
</div>
`;

exports[`First winner creation snapshot for a player 1`] = `
<div
  class="resultList"
>
  <p>
    You 
    <b>
      win
    </b>
    .
  </p>
  <ul
    class="weaponsList"
  >
    <li>
      <img
        alt="scissors"
        src="./pics/scissors.png"
        width="100"
      />
    </li>
    <li>
      VS.
    </li>
    <li>
      <img
        alt="paper"
        src="./pics/paper.png"
        width="100"
      />
    </li>
  </ul>
</div>
`;

exports[`Image creation snapshot 1`] = `
<img
  alt="undefined"
  src="./pics/undefined.png"
  width="100"
/>
`;

exports[`Second winner creation snapshot 1`] = `
<div
  class="resultList"
>
  <p>
    <b>
      scissors
    </b>
     wins.
  </p>
  <ul
    class="weaponsList"
  >
    <li>
      <img
        alt="paper"
        src="./pics/paper.png"
        width="100"
      />
    </li>
    <li>
      VS.
    </li>
    <li>
      <img
        alt="scissors"
        src="./pics/scissors.png"
        width="100"
      />
    </li>
  </ul>
</div>
`;

exports[`Second winner creation snapshot for a player 1`] = `
<div
  class="resultList"
>
  <p>
    You 
    <b>
      lose
    </b>
    .
  </p>
  <ul
    class="weaponsList"
  >
    <li>
      <img
        alt="paper"
        src="./pics/paper.png"
        width="100"
      />
    </li>
    <li>
      VS.
    </li>
    <li>
      <img
        alt="scissors"
        src="./pics/scissors.png"
        width="100"
      />
    </li>
  </ul>
</div>
`;

exports[`Spock snapshot 1`] = `
<img
  alt="You have been Spocked"
  class="spocked"
  src="./pics/spocked.jpg"
  title="You have been Spocked"
/>
`;

exports[`Weapons creation snapshot 1`] = `
<ul
  class="weaponsList"
>
  <li>
    <img
      alt="firstWeap"
      class="weapon"
      src="/pics/firstWeap.png"
      width="100"
    />
  </li>
  <li>
    <img
      alt="secondWeap"
      class="weapon"
      src="/pics/secondWeap.png"
      width="100"
    />
  </li>
</ul>
`;
